File: tools/lib/artifact-inputs.js
/**
 * Campaign-artifact input fingerprinting (§12h follow-up work, Phase 1).
 *
 * A generated artifact in data/ records WHAT it was computed from: an
 * `inputs` block mapping each input file (and the generator script itself)
 * to a sha256 of its bytes. The freshness gate
 * (tools/verify/artifact-freshness.js) re-hashes those paths on every
 * `npm run check` and FAILS when any input moved without the artifact being
 * regenerated — which is what makes a hand-stale campaign number
 * structurally impossible without paying the generator's runtime per check.
 *
 * Convention (shared with the Python twin tools/fit/python/artifact_inputs.py):
 *   "inputs": {
 *     "generator": "node tools/explore/cassini-moontilt.js --write",
 *     "files": { "<repo-relative path>": "sha256:<hex>", ... }
 *   }
 * The generator script itself MUST be one of the files — an edited generator
 * with an un-regenerated artifact is stale by definition.
 */

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..', '..');

/** sha256 of a repo-relative file, formatted for the inputs block. */
function hashFile(relPath) {
  const bytes = fs.readFileSync(path.join(ROOT, relPath));
  return 'sha256:' + crypto.createHash('sha256').update(bytes).digest('hex');
}

/** Build the `inputs` block for a generator invocation + its input files. */
function buildInputsBlock(generatorCommand, relPaths) {
  const files = {};
  for (const rel of [...relPaths].sort()) files[rel] = hashFile(rel);
  return { generator: generatorCommand, files };
}

module.exports = { ROOT, hashFile, buildInputsBlock };

File: tools/fit/python/artifact_inputs.py
"""Campaign-artifact input fingerprinting — Python twin of
tools/lib/artifact-inputs.js (same convention, same gate).

    "inputs": {
      "generator": "python tools/fit/python/eval_precession_physical.py --write",
      "files": { "<repo-relative path>": "sha256:<hex>", ... }
    }

The generator script itself MUST be listed — an edited generator with an
un-regenerated artifact is stale by definition. The freshness gate
(tools/verify/artifact-freshness.js, in `npm run check`) re-hashes every
listed path and fails on any mismatch.
"""

import hashlib
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent.parent.parent


def hash_file(rel_path: str) -> str:
    digest = hashlib.sha256((ROOT / rel_path).read_bytes()).hexdigest()
    return f'sha256:{digest}'


def build_inputs_block(generator_command: str, rel_paths) -> dict:
    return {
        'generator': generator_command,
        'files': {rel: hash_file(rel) for rel in sorted(rel_paths)},
    }
